File: lib/pieces/general.rb
require_relative "./piece"

class General < Piece

  def get_moves
    directions.each do |direction|
      target_square = starting_square.send direction
      next unless target_square && in_palace?(target_square)
      target_piece = @position.find target_square.coordinates
      next if target_piece && target_piece.color == @color
      add_move target_square.coordinates
    end
  end

  def sees_other?
    next_square = starting_square.send forward
    while next_square
      current_square = next_square
      target_piece = @position.find current_square.coordinates
      if target_piece
        return true if target_piece.class == General
        break
      end
      next_square = current_square.send forward
    end
    false
  end
end

File: src/general.js
if (typeof window === 'undefined') {
  var Piece = require('./piece.js');
  module.exports = General;
}

function General(color) {
  var self = new Piece(color);

  self.type = 'General';

  self.getMoves = function(position) {
    var moves = [];
    var palace = position.BOARD.palaces[this.color];
    var directions = ['left', 'right', 'up', 'down'];
    for(var dir_index = 0; dir_index < 4; dir_index += 1) {
      direction = directions[dir_index];
      target_square = this.square[direction];
      if (!target_square || !palace[target_square.coordinates]) {
        continue;
      }
      if (!position[target_square.coordinates] ||
        position[target_square.coordinates].color !== this.color) {
        moves.push(target_square.coordinates);
      }
    }
    return moves;
  };

  return self;
}
